src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -110,44 +121,126 @@ protected override void Dispose(bool disposing)
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void StartRequest(RequestCompletionSource state)
         {
-            HttpResponseMessage responseMessage = null;
+            CurlResponseMessage responseMessage = null;
             SafeCurlHandle requestHandle = null;
+            GCHandle requestStateHandle = new GCHandle();
+            IntPtr headerPtr = IntPtr.Zero;
+            bool referenceTaken = false;
+
+#if DEBUG
+            if (Logging.On) Logging.Enter(Logging.Http, this, "StartRequest", state.RequestMessage);
+#endif
 
             if (state.CancellationToken.IsCancellationRequested)
             {
-                state.TrySetCanceled();
+                state.TrySetCanceled(state.CancellationToken);
                 return;
             }
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+            var cancellationTokenRegistration = state.CancellationToken.Register(
+                s =>
+                {
+                    var rcs = (RequestCompletionSource)s;
+                    rcs.TrySetCanceled(rcs.CancellationToken);
```

Comment:
Sorry.. I had misunderstood the earlier comment (https://github.com/dotnet/corefx/pull/2473#discussion_r35552624 ). I think it is better to just get rid of the registration itself. Do let me know if you see an issue with that